cpp/src/arrow/python/serialize.cc

```diff
@@ -63,297 +70,195 @@ class SequenceBuilder {
       : pool_(pool),
         types_(::arrow::int8(), pool),
         offsets_(::arrow::int32(), pool),
-        nones_(pool),
-        bools_(::arrow::boolean(), pool),
-        ints_(::arrow::int64(), pool),
-        py2_ints_(::arrow::int64(), pool),
-        bytes_(::arrow::binary(), pool),
-        strings_(pool),
-        half_floats_(::arrow::float16(), pool),
-        floats_(::arrow::float32(), pool),
-        doubles_(::arrow::float64(), pool),
-        date64s_(::arrow::date64(), pool),
-        tensor_indices_(::arrow::int32(), pool),
-        ndarray_indices_(::arrow::int32(), pool),
-        buffer_indices_(::arrow::int32(), pool),
-        list_offsets_({0}),
-        tuple_offsets_({0}),
-        dict_offsets_({0}),
-        set_offsets_({0}) {}
+        type_map_(PythonType::MAX) {
+    builder_.reset(new DenseUnionBuilder(pool));
+  }
 
   // Appending a none to the sequence
-  Status AppendNone() {
-    RETURN_NOT_OK(offsets_.Append(0));
-    RETURN_NOT_OK(types_.Append(0));
-    return nones_.AppendNull();
-  }
+  Status AppendNone() { return builder_->AppendNull(); }
 
-  Status Update(int64_t offset, int8_t* tag) {
-    if (*tag == -1) {
-      *tag = num_tags_++;
-    }
+  template <typename BuilderType>
+  Status Update(BuilderType* child_builder, int8_t tag) {
     int32_t offset32 = -1;
-    RETURN_NOT_OK(internal::CastSize(offset, &offset32));
+    RETURN_NOT_OK(internal::CastSize(child_builder->length(), &offset32));
     DCHECK_GE(offset32, 0);
-    RETURN_NOT_OK(offsets_.Append(offset32));
-    RETURN_NOT_OK(types_.Append(*tag));
-    return nones_.Append(true);
+    return builder_->Append(tag, offset32);
+  }
+
+  template <typename BuilderType, typename MakeBuilderFn>
+  Status CreateAndUpdate(std::shared_ptr<BuilderType>* child_builder, int8_t tag,
+                         MakeBuilderFn make_builder) {
+    if (!*child_builder) {
+      child_builder->reset(make_builder());
+      type_map_[tag] = builder_->AppendChild(*child_builder, std::to_string(tag));
```

Comment:
Sidenote: `std::to_string` and `std::stoi` are locale-dependent. Hopefully it shouldn't be a problem here, as we're only handling relatively small integers...

Reply:
ok, let me know if there is a better way to do this.

Reply:
Not an easy way. I suggest we leave it like that for now, perhaps add a comment?

Reply:
comments added